src/lib/reasoners/bitv.ml

```diff
@@ -289,40 +315,557 @@ module Shostak(X : ALIEN) = struct
+        match X.term_extract r with
+        | Some t, _ ->
+          begin match bit_of_expr t with
+            | Some t_b ->
+              let nctx = ctx' @ ctx in
+              begin match prec with
+                | Some p_b when Bool.equal t_b p_b ->
+                  if ncnt > size then { bv = I_Cte (bitf t_b); sz = sz}, nctx
+                  else
+                    aux
+                      ?prec ncnt (sz + 1) (get_nth_eq_b bv_t cnt t_b :: nctx)
+                | Some p_b ->
+                  let t1 = { bv = I_Cte (bitf p_b); sz = sz} in
+                  if ncnt > size then t1, nctx
+                  else
+                    let t2, nctx =
+                      aux ~prec:t_b ncnt 1 (get_nth_eq_b bv_t cnt t_b :: nctx)
+                    in
+                    mk_i_comp t2 t1 (sz + t2.sz), nctx
+                | None ->
+                  assert (ncnt <= size);
+                  let t, nctx =
+                    aux ~prec:t_b ncnt 1 (get_nth_eq_b bv_t cnt t_b :: nctx)
+                  in
+                  t, nctx
+              end
+            | None ->
+              let rv_t = bit_e_neg neg val_t in
+              let bvv = E.fresh_name (Ty.Tbitv 1) in
+              let eq = E.mk_eq ~iff:false bvv rv_t in
+              let r = X.term_embed bvv in
+              let nctx = eq :: ctx' @ ctx in
+              let t1 = { bv = I_Other r; sz = 1 } in
+              if ncnt >= size then t1, nctx
+              else
+                let t2, nctx = aux ncnt 0 nctx in
+                mk_i_comp t2 t1 (t2.sz + 1), nctx
+          end
+        | None, _ ->
+          let rv_t = bit_e_neg neg val_t in
+          let r, ctx' = X.make rv_t in
+          let nctx = ctx' @ ctx in
+          begin match prec with
+            | Some b ->
+              let t1 =
+                mk_i_comp
+                  { bv = I_Other r; sz = 1}
+                  { bv = I_Cte (bitf b); sz }
+                  (sz + 1)
+              in
+              if ncnt > size then t1, nctx
+              else
+                let t2, nctx =
+                  aux ncnt 0 (mk_nthbv_eq bv_t cnt rv_t :: nctx)
+                in
+                let t = mk_i_comp t2 t1 (sz + t2.sz + 1) in
+                t, nctx
+            | None ->
+              let t1 = { bv = I_Other r; sz = 1} in
+              if ncnt > size then t1, nctx
+              else
+                let t2, nctx =
+                  aux ncnt 0 (mk_nthbv_eq bv_t cnt rv_t :: nctx)
+                in
+                let t = mk_i_comp t2 t1 (t2.sz + 1) in
+                t, nctx
+          end
+      in
+      aux 0 0 ctx
+
+    let bvor_ctes ~is_and b1 sz1 b2 sz2 =
+      let bres = if is_and then b1 && b2 else b1 || b2 in
+      if sz1 > sz2 then
+        {bv = Cte bres; sz = sz2}, [{ bv = Cte b1; sz = sz1 - sz2 }], []
+      else if sz1 < sz2 then
+        {bv = Cte bres; sz = sz1}, [], [{ bv = Cte b1; sz = sz2 - sz1 }]
+      else {bv = Cte bres; sz = sz1}, [], []
```

Comment:
Not sure I understand, all the pairs of arguments to `bvor` have the same size here?

Ah wait this is the case that performs the split for `or`, nevermind.

Hm. I have not dug deep enough into the details, but would it be possible to extract the part of the function responsible for extraction, so that we can write something like

```
let h1, tl1 = extract_first_n bv1 (min sz1 sz2) in
let h2, tl2 = extract_first_n bv2 (min sz1 sz2) in
let h, ctx = bvor_uniform h1 h2 in
let t, ctx = aux tl1 tl2 ctx in
h :: t, ctx
```

and isolate the repeated `if-then-else` patterns? This would make it more readable & less error-prone.

Reply:
True, it's a lot simpler and more elegant, thanks! it's done